New debugger (#116)
madewithkoji/docs: src/asciidoc/docs/develop/testingTemplates.adoc

The following contexts are available for selection.

==== Default

Selecting *Default* shows you a preview of the running template.
Events associated with the default context are displayed in an event pane.
See <<_viewing_events>>.

In the *Default* context, you can see what happens when the template asks the platform for things like <<withkoji-koji-iap-package#, in-app purchases>> or <<withkoji-koji-auth-sdk#, auth grants>>.

==== Remix

Selecting *Remix* displays a special remixing page, if you defined one in your template.
Otherwise, it displays the editable elements of your template and allows you to change them.
A separate pane displays the *VCC Values* as they appear in the code for the template.
Events associated with the *Remix* context are displayed in an event pane.
See <<_viewing_events>>.

*Remix* gives you a dynamic view of remix events, for a better understanding of how VCC data is changing over time in response to user actions.

==== Thumbnail

Selecting the *Thumbnail* shows you exactly how your template will look if you share it on social media.
See <<rich-preview-image#>>.

*Thumbnail* shows previews in both square and rectangular format.

==== Admin

> If the admin context is enabled, selecting *Admin* displays the admin page.

Reviewer: Here's another example of redundant phrasing (admin). One possible fix (possibly not the best):
```suggestion
Select *Admin* to debug the admin context, if you defined one in your template.
```

Events associated with the admin context are displayed in an events pane.
See <<_viewing_events>>.

==== About

If the about context is enabled, clicking *About* displays the page that appears when the user clicks *Remix this Koji*.
Events associated with the about context are displayed in an events pane.
See <<_viewing_events>>.

==== Sticker

The *Sticker* context shows you how your template behaves when it's embedded inside another template.
Events associated with the sticker context are displayed in an events pane.
See <<_viewing_events>>.

==== Dynamic Receipt

If a dynamic receipt is enabled, a page is displayed for editing it.
See <<withkoji-koji-iap-package#, in-app purchases>>.

> === Viewing events

Reviewer: This section also could use some tightening up. I believe this functionality is very similar to the "viewing messages", except real-time/in-production, so maybe we can use similar text.

Author: I don't quite follow how this is similar to the "viewing messages" section. I did change the wording to make it more similar to wording for the other contexts. See what you think.

The template generates events that are sent to the Koji platform, and the Koji platform generates events that are sent to the template.
In most cases, selecting one of the contexts in the debugger displays an events pane which shows the events associated with that context.
An events pane allows you to see in real time the events that are generated as you interact with the template in the selected context.

> Events generated by the template are listed on the right side of the events pane.

Reviewer: Suggest removing these two sentences about the side of the pane. The UI could change, and it wouldn't affect the way you use it. The events indicate how they were generated.

Events generated by the Koji platform are listed on the left side.

The info for each event includes:

* The type of event – PLATFORM or TEMPLATE
* The name of the event – For example: *Checked*
* Properties set by the event (optional) – This may include descriptions of the properties.
* New values for the properties (optional) – The new values for the properties set by the event.

If you open a template and select the *Remix* context, here's an example of some events that you might see in the events pane:

* TEMPLATE +
Ready

* PLATFORM +
Changing remix state +
VALUE +
Remixing

* TEMPLATE +
Set value +
PATH +
+++[general, title]+++ +
NEW VALUE +
+++""+++

If you click in an editable field on the remixing page, a *Clicked* event will appear in the event pane.
For example:

> * TEMPLATE +

Reviewer: I'm not sure about using a bulleted list to format these example events. Maybe there's another way to call them out? A code or an example block, perhaps?

Clicked +
POSITION +
(131, 583)

If you then change the value of the field you clicked, a *Set Value* event will appear.
For example:

* TEMPLATE +
Set value +
PATH +
+++[general, goal]+++ +
NEW VALUE +
"21.00" +
TYPE +
Skip update

=== Info and Features

You can view and modify environment info for your template, as well as settings and data for implemented features.

> If you click on a feature that is not implemented, a message will be displayed saying that no data was found for the feature.

Reviewer: ```suggestion
If a feature is not implemented, a message indicates that no data was found for it.
```
Bit shorter, and I don't think you need to click to see the message.

The *Info and Features pane* includes the following panels.

==== Environment info

> The *Environment info* panel displays your templates's *App Id*, *Frame URL*, and *VCC editor source*.

Reviewer: ```suggestion
The *Environment info* panel displays your template's *App Id*, *Frame URL*, and *VCC editor source*.
```

> You can click *Modify* to edit the frame URL.

Reviewer: I actually don’t see a place to edit the Frame URL, so I don’t know if I’m missing something or if the UI has changed. We might need some tech info from Sean.

Author: Hmm... maybe it did change. Or maybe I was seeing things.

The *Environment info* panel also allows you to edit the *VCC editor source*.
When you publish a template, the VCC editor data defines the VCCs available in the template.
So when the template calls something like `InstantRemixing.onPresentControl(['general', 'photo'])`, the editor looks in the data for the `general.photo` path and knows to show the image VCC.
If you're working on a new version of a template that uses different VCCs from the currently published template, and you're using the debugger to override the URL and debug something running locally or running in the web IDE, you can edit the *VCC editor source* to override the live data.

You edit the *VCC editor source* by clicking *Modify* and selecting *Override and use different VCC editor data*.
This displays the data in JSON format.
You can directly edit the JSON code to override the live values.

> For example, suppose the VCC editor data for a template reads as follows:

Reviewer: The VCC editor data seems like too much detail for this page. Someone who’s is at this point should already know about VCC data. Your example use case – switching from an `image` type to a `video` type – is very useful and should be the highlighted part. Consider removing the rest of the JSON and jumping straight into the use case.

        "type": "image"
      },
      {
        "key": "newMessageBackground",
        "name": "New Message Background",
        "type": "image"
      },
      {
        "key": "photo",
        "name": "Photo",
        "type": "image"
      },
      {
        "key": "video",
        "name": "Video",
        "type": "video"
      },
      {
        "key": "price",
        "name": "Price Per Message",
        "type": "text"
      }
    ],
    "source": "general.json"
  }
]
----

The following portion of the JSON data tells Koji that the template will use a profile image and, when remixing, the prompt for that image will be "Profile Image":

[source, json]
----
{
    "key": "profileSrc",
    "name": "Profile Image",
    "type": "image"
},
----

Suppose you're working on a new version of the template that uses a profile video instead of a profile image.
You can override the live data by changing `"type": "image"` to `"type": "video"`, and by changing `"name": "Profile Image",` to `"name": "Profile Video",`.
Now, when working inside the debugger, the remixer will show a video VCC instead of an image VCC, and the prompt for changing the video will be "Profile Video".

> Without using the override, you would need to update the VCC definition and republish in order to preview the change.

Reviewer: Needs something simpler and more straightforward to explain why this is important. Here's one suggestion. Also, consider making this point before you describe the example. It might flow better.

```suggestion
Overriding the VCC data enables you to test the new VCC definition without republishing the template.
```

==== Database

If your template uses the Koji database, the database views will be displayed in a panel, and the records for each view can be displayed.
*Database* also gives you an easy way to delete records.

==== In-app purchase receipts

> If *in-app purchases* is implemented in your template, the *in-app purchase receipts* panel allows you to view and delete receipts.

Reviewer: ```suggestion
If in-app purchases are implemented in your template, the *In-app purchase receipts* panel allows you to view and delete receipts.
```

==== Auth grants

> If your template has any auth grants, *Auth grants* displays them.

Reviewer: Consider making this section parallel with the previous section.
